source/common/network/udp_listener_impl.cc

```diff
@@ -125,5 +125,49 @@ Api::IoCallUint64Result UdpListenerImpl::flush() {
   return cb_.udpPacketWriter().flush();
 }
 
+void UdpListenerImpl::activateRead() { file_event_->activate(Event::FileReadyType::Read); }
+
+UdpListenerWorkerRouterImpl::UdpListenerWorkerRouterImpl(uint32_t concurrency)
+    : workers_(concurrency) {}
+
+void UdpListenerWorkerRouterImpl::registerWorker(UdpListenerCallbacks& listener) {
+  absl::WriterMutexLock lock(&mutex_);
+
+  ASSERT(listener.workerIndex() < workers_.size());
+  workers_.at(listener.workerIndex()) = &listener;
+}
+
+void UdpListenerWorkerRouterImpl::unregisterWorker(UdpListenerCallbacks& listener) {
+  absl::WriterMutexLock lock(&mutex_);
+
+  ASSERT(workers_.at(listener.workerIndex()) == &listener);
+  workers_.at(listener.workerIndex()) = nullptr;
+}
+
+void UdpListenerWorkerRouterImpl::deliver(UdpListenerCallbacks& current, UdpRecvData&& data) {
+  absl::ReaderMutexLock lock(&mutex_);
+
```

Comment:
Instead of calling UdpListenerWorkerRouter::deliver() from UdpListenerCallbacks and then call UdpListenerCallbacks::destination(), can you call destination() from UdpListenerCallbacks and then call UdpListenerWorkerRouter::deliver() with UdpRecvData&& if the packet doesn't belong to the current UdpListenerCallbacks instance? All UdpListenerCallbacks::destination() needs from UdpListenerWorkerRouter is the number of workers which is a const since start up.

Doing so reduce a level of indirection, and avoid lock if the packet doesn't need to be delivered to other workers.

Reply:
That's a really good point about removing the need to lock when the packet isn't getting redirected.